qadence_libs/qinfo_tools/qng.py

```diff
@@ -0,0 +1,182 @@
+
+import torch
+from qadence import QuantumCircuit
+from torch.optim.optimizer import Optimizer, required
+
+from qadence_libs.qinfo_tools.qfi import get_quantum_fisher, get_quantum_fisher_spsa
+
+
+class QuantumNaturalGradient(Optimizer):
+    """Implements the Quantum Natural Gradient Algorithm.
+
+    WARNING: This class implements the exact QNG optimizer, which is very inefficient
+    both in time and memory as it calculates the exact Quantum Fisher Information of
+    the circuit at every iteration. Therefore, it is not meant to be run with medium
+    to large circuits. Other approximations such as the QNG-SPSA optimizer are much
+    more efficient while retaining good performance.
+    """
+
+    def __init__(
+        self,
+        params: tuple | torch.Tensor,
+        circuit: QuantumCircuit = required,
+        lr: float = required,
+        beta: float = 10e-3,
+    ):
+        """
+        Args:
+
+            params (tuple | torch.Tensor): Variational parameters to be updated
+            circuit (QuantumCircuit): Quantum circuit. Needed to compute the QFI matrix
+            lr (float): Learning rate.
+            beta (float):
+                Shift applied to the QFI matrix before inversion to ensure numerical stability.
+                Defaults to 10e-3.
+        """
+
+        if 0.0 > lr:
+            raise ValueError(f"Invalid learning rate: {lr}")
+        if 0.0 >= beta:
+            raise ValueError(f"Invalid beta value: {beta}")
+        if not isinstance(circuit, QuantumCircuit):
+            raise ValueError(f"The circuit should be an instance of {type(QuantumCircuit)}")
+
+        defaults = dict(circuit=circuit, lr=lr, beta=beta)
+        super(QuantumNaturalGradient, self).__init__(params, defaults)
+
+    def __setstate__(self, state):  # type: ignore
+        super().__setstate__(state)
+
+    def step(self, closure: Callable | None = None) -> torch.Tensor:
+        """Performs a single optimization step of the QNG algorithm.
+
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        for group in self.param_groups:
+            grad_vec = torch.Tensor([v.grad.data for v in group["params"] if v.requires_grad])
+
+            # Calculate the EXACT metric tensor
+            metric_tensor = 0.25 * get_quantum_fisher(
+                group["circuit"],
+                vparams_values=group["params"],
+            )
+
+            with torch.no_grad():
+                # Finite shift the metric tensor to avoid problems when inverting
+                metric_tensor = metric_tensor + group["beta"] * torch.eye(len(grad_vec))
+
+                # Get transformed gradient vector
+                metric_tensor_inv = torch.linalg.inv(metric_tensor)
+                transf_grad = torch.matmul(metric_tensor_inv, grad_vec)
```

Comment:
So I came to check this after your presentation, it might be worth changing this to solve a linear system instead. Since you seem to be doing $x=A^{-1}b$ for `x=transf_grad`, typically the numerics of solving $Ax=b$ are much more efficient than those of computing $A^{-1}$.

https://www.johndcook.com/blog/2010/01/19/dont-invert-that-matrix/

Reply:
Same thing for the usage of `torch.linalg.pinv()`

https://pytorch.org/docs/stable/generated/torch.linalg.pinv.html
https://pytorch.org/docs/stable/generated/torch.linalg.lstsq.html#torch.linalg.lstsq

Reply:
That's a great suggestion, thanks!

Reply:
This has been implemented.